fuel-vm/src/interpreter/memory.rs

```diff
@@ -19,23 +22,324 @@ use fuel_types::{
+    #[derivative(Debug(format_with = "fmt_truncated_hex::<16>"))]
+    heap: Vec<u8>,
+    /// Lowest allowed heap address, i.e. hp register value.
+    /// This is needed since we can allocate extra heap for performance reasons.
+    hp: usize,
+}
+
+impl Default for Memory {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Memory {
+    /// Create a new VM memory.
+    pub fn new() -> Self {
+        Self {
+            stack: Vec::new(),
+            heap: Vec::new(),
+            hp: MEM_SIZE,
+        }
+    }
+
+    /// Offset of the heap section
+    fn heap_offset(&self) -> usize {
+        MEM_SIZE.saturating_sub(self.heap.len())
+    }
+
+    /// Returns a linear memory representation where stack is at the beginning and heap is
+    /// at the end.
+    pub fn into_linear_memory(self) -> Vec<u8> {
+        let uninit_memory_size = MEM_SIZE
+            .saturating_sub(self.stack.len())
+            .saturating_sub(self.heap.len());
+        let uninit_memory = vec![0u8; uninit_memory_size];
+        let mut memory = self.stack;
+        memory.extend(uninit_memory);
+        memory.extend(self.heap);
+        memory
+    }
+
+    /// Grows the stack to be at least `new_sp` bytes.
+    pub fn grow_stack(&mut self, hp: Reg<HP>, new_sp: Word) -> Result<(), PanicReason> {
+        let new_sp_word = new_sp.min(MEM_SIZE as Word);
+        #[allow(clippy::cast_possible_truncation)] // Safety: MEM_SIZE is usize
+        let new_sp = new_sp_word as usize;
+        if new_sp > self.stack.len() {
+            if new_sp_word > *hp {
+                return Err(PanicReason::MemoryGrowthOverlap)
+            }
+
+            self.stack.resize(new_sp, 0);
+        }
+        Ok(())
+    }
+
+    /// Grows the heap to be at least `new_hp` bytes.
+    pub fn grow_heap(&mut self, sp: Reg<SP>, new_hp: Word) -> Result<(), PanicReason> {
+        let new_hp_word = new_hp.min(MEM_SIZE as Word);
+        #[allow(clippy::cast_possible_truncation)] // Safety: MEM_SIZE is usize
+        let new_hp = new_hp_word as usize;
+
+        if new_hp_word < *sp {
+            return Err(PanicReason::MemoryGrowthOverlap)
+        }
+
+        // Expand the heap allocation
+        reverse_resize_at_least(&mut self.heap, MEM_SIZE - new_hp);
+        self.hp = new_hp;
+
+        // If heap enters region where stack has been, truncate the stack
+        self.stack.truncate(new_hp);
+
+        Ok(())
+    }
+
+    /// Verify that the memory range is accessble and return it as a range.
+    pub fn verify<A: ToAddr, B: ToAddr>(
+        &self,
+        addr: A,
+        count: B,
+    ) -> Result<MemoryRange, PanicReason> {
+        let start = addr.to_addr()?;
+        let len = count.to_addr()?;
+        let end = start.saturating_add(len);
+        if end > MEM_SIZE {
+            return Err(PanicReason::MemoryOverflow)
+        }
+
+        if end <= self.stack.len() || (start >= self.hp && start >= self.stack.len()) {
```

Comment:
I'm just curious why we need `start >= self.stack.len()` if `hp >= sp`? For the cases where `sp < self.stack.len()`? If yes, then it is a bug in the memory implementation(on our side), and we need to return something else than `UninitalizedMemoryAccess`

Reply:
Good catch. I added the check before I realized that we have to truncate the stack if heap extends on top of it. WIll remove.

Reply:
Removed in e8f5b451